### PR TITLE
Improve performance when tagged debug output is enabled.

### PR DESCRIPTION
**Pull request by ywkaras**
This is for issue https://github.com/apache/trafficserver/issues/6150 .

### CODE REVIEW

**Comment by ywkaras** [comment]
This looks promising.  Repeating David's test, I get:
```
finished in 60.05s, 356.00 req/s, 146.02KB/s
```
with debug output disabled, and:
```
finished in 60.03s, 261.48 req/s, 107.25KB/s
```
with debug output enabled (testing in a Docker container on my MacBook).  I did the second test with TS stdout/err redirected to a file, which makes it much faster than if output goes to /dev/tty (not sure how David did the original test).

The next step is to add some automated testing.

This eventually could be expanded to return a pointer to DiagEnabled for the tag.  A new debug output function could be added to take this pointer instead of the tag (avoiding the unordered_map lookup for each debug call).  The TS API could be modified with a similar capability.

**Comment by ywkaras** [comment]
This passed the autest PR check but I'm seeing that the schedule_on_pool test is flakey in my dev env.

**Comment by ywkaras** [comment]
I did more comprehensive testing, using a URL with a cacheable response.  The results indicate this change at best makes a tiny dent in the problem:
```
Test URL (cacheable): http://httpbin.org/cache/600

load command:
h2load --h1 -t 40 -D 60 -n 10000 -c 40 http://127.0.0.1:8080

run ts:
./traffic_server >| /tmp/out 2>&1

With code change:

diags.debug.enabled = 0

finished in 60.03s, 27399.58 req/s, 16.67MB/s

diags.debug.enabled = 1

finished in 60.02s, 343.45 req/s, 214.32KB/s

With commit before change ( 080889f68701c994826e2e5781d2564b4ed6733d ):

diags.debug.enabled = 0

finished in 60.03s, 26953.67 req/s, 16.40MB/s

diags.debug.enabled = 1

finished in 60.03s, 340.07 req/s, 212.21KB/s

run ts:
./traffic_server > /dev/null 2> /dev/null

With code change:

diags.debug.enabled = 0

finished in 60.03s, 26498.60 req/s, 16.12MB/s

diags.debug.enabled = 1

finished in 60.05s, 906.88 req/s, 565.92KB/s

With commit before change:

diags.debug.enabled = 0

finished in 60.04s, 26917.62 req/s, 16.37MB/s

diags.debug.enabled = 1

finished in 60.07s, 583.85 req/s, 364.34KB/s

run ts (output goes to /dev/tty):
./traffic_server

With code change:

diags.debug.enabled = 0

finished in 60.10s, 27137.08 req/s, 16.54MB/s

diags.debug.enabled = 1

finished in 60.03s, 73.83 req/s, 45.99KB/s
```

First thing to note is that the time it takes the OS to handle the trace output seems to be a big factor.  Running TS without my change, with diags.debug.enabled = 1, the performance is 340 req/s if the trace output is going to a disk file, and 584 req/s if the trace output is going to /dev/null.

With trace output going to /dev/null, with diags.debug.enabled = 1, the performance is 907 req/s with my code changes, versus 584 reqs/s without my code change.  But performance is about 27K req/s with diags.debug.enabled = 0.

Something to keep in mind is that the size of the response for the target URL in this test is very small.  The amount of I/O to do the trace output for each transaction is much bigger than the amount of I/O to read in the object.  If I did this testing with URL with a large response, I think the effect of turning on tagged debug output would be much less.

**Comment by ywkaras** [comment]
It looks like the schedule_on_pool Au test is flakey (fails intermittently) even without my code change.

**Comment by zwoop** [comment]
[approve ci]

**Comment by ywkaras** [comment]
I added unit tests (and more comments).  I fixed a bug that I found as a result.  I re-ran the performance tests, but with diags.debug.tags set to `do_not_match_anything` rather than the default `http|dns`.  This better isolates the overhead from simply enabling diags.debug.enabled from the actual overhead of doing the text output.  These test results (below) indicate that my code changes are highly desirable (17800 req/s versus 8664 req/s).

There is a big difference between the nominally equivalent result here using the baseline commit and diags.debug.enabled set to zero, 21267 req/s, and the previous result, 26954 req/s.  I have no explanation for this.

The value of this change should not be overestimated.  If you have many tags or a high volume tags enabled, there will still be a major performance hit.  I've heard there is a way to limit debug output to a particular client IP address, but I can't find information on it in the documentation.

As I mentioned before, the small size of the response body in these tests I think exaggerate both the size of the problem and the value of the solution.

```
Test URL (cacheable): http://httpbin.org/cache/600

load command:
h2load --h1 -t 40 -D 60 -n 10000 -c 40 http://127.0.0.1:8080

run ts:
./traffic_server >| /tmp/out 2>&1

diags.debug.tags = do_not_match_anything

With code change:

diags.debug.enabled = 0

finished in 60.04s, 22088.98 req/s, 13.44MB/s

diags.debug.enabled = 1

finished in 60.08s, 17800.40 req/s, 10.85MB/s

With commit before change ( 080889f68701c994826e2e5781d2564b4ed6733d ):

diags.debug.enabled = 0

finished in 60.05s, 21266.90 req/s, 12.96MB/s

diags.debug.enabled = 1

finished in 60.06s, 8664.03 req/s, 5.28MB/s
```

**Comment by ywkaras** [comment]
[approve ci autest]

**Comment by ywkaras** [comment]
[approve ci autest]

**Comment by ywkaras** [comment]
I ran the load test on TS without my change, with a (cached) HTTP response that's about 4MB.  The results are 4.6 req/s, 19 MB/s, with or without debug enabled.  The throughput in bytes per second goes way up (by about half), but the request per second goes way down.  So the importance of debug overhead clearly depends on average response size.

**Comment by ywkaras** [comment]
When I wrote this code, I did not realize there was a pthreads implementation of one-writer-mulitple-reader locks.  I changed this code to use the pthread implementation rather than mine:  https://github.com/ywkaras/trafficserver/tree/FasterDiagsPthread .  I re-ran the speed test.  The result was 10114.32 req/s, versus 17800.40 req/s with my version of the lock.  This indicates that my lock implementation is preferable for this use-case.  It also indicates that the reduction in PCRE calls is less help in improving performance than reducing locking overhead.

File: src/tscore/OneWriterMultiReader.h [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+#pragma once
+
+#include <atomic>
+#include <mutex>
+#include <condition_variable>
+
+namespace ts
+{
+class OneWriterMultiReader
```
[the remark this answers is not in the document]
Comment (ywkaras) [reply src/tscore/OneWriterMultiReader.h]: Ooops I didn't realize this made it into 17.  But std::shared_ptr is probably based on pthread_rwlock_t, in which case there is a performance advantage to using my code:  https://github.com/apache/trafficserver/pull/6255#issuecomment-580983937 .

**Comment by ywkaras** [comment]
[approve ci FreeBSD]

**Comment by ywkaras** [comment]
[approve ci docs]

**Comment by ywkaras** [comment]
[approve ci autest]

**Comment by ywkaras** [comment]
We decided, nah.

**Comment by repodude** [comment]
Reason-to-be anecdote:   I work on an ATS plugin whose outcome is dependent on many dozens of factors such that it can be very difficult to predict outcome.  I use TSDEBUG() logging to output decision points and logic flow.   It is critical during development and extremely helpful production integration testing to verify correct output.    Unfortunately, at production volumes this is not available to me.   Turning on debug for just a small amount of output can drive my machine load average from 16 to 64.   Sometimes issues I'm tracking only come from some weird external case that I don't yet know all of the intricate conditions.   it would be very useful to have TSDEBUG() logging functionality working such that it is a viable option with production level traffic.

**Comment by ywkaras** [comment]
@repodude I've rebased https://github.com/ywkaras/trafficserver/tree/FasterDiags to the tip of master, if you want to give it a try.  (There were no conflicts.). Have you considered just changing your debug printfs to standard printfs in your test build?

**Pull request closed without merging**